src/core/memory.d

```diff
@@ -809,3 +809,41 @@ struct GC
         gc_runFinalizers( segment );
     }
 }
+
+/// Pure variant of `malloc` that doesn't affect global variable `errno`.
+void* pureMalloc(size_t size) pure
+{
+    const errno = _pureGetErrno();
+    void* ret = _pureMalloc(size);
+    cast(void)_pureSetErrno(errno);
+    return ret;
+}
+
+/// Pure variant of `calloc` that doesn't affect global variable `errno`.
+void* pureCalloc(size_t nmemb, size_t size) pure
+{
+    const errno = _pureGetErrno();
+    void* ret = _pureCalloc(nmemb, size);
+    cast(void)_pureSetErrno(errno);
+    return ret;
+}
+
+/// Pure variant of `realloc` that doesn't affect global variable `errno`.
+void* pureRealloc(void* ptr, size_t size) pure
+{
+    const errno = _pureGetErrno();
+    void* ret = _pureRealloc(ptr, size);
+    cast(void)_pureSetErrno(errno);
+    return ret;
+}
+
+// locally purified for internal use here only
+extern (C) private pure
+{
+    pragma(mangle, "getErrno") int _pureGetErrno();
+    pragma(mangle, "setErrno") int _pureSetErrno(int);
+
+    pragma(mangle, "malloc") void* _pureMalloc(size_t);
+    pragma(mangle, "calloc") void* _pureCalloc(size_t nmemb, size_t size);
+    pragma(mangle, "realloc") void* _pureRealloc(void* ptr, size_t size);
+}
```

Comment:
please add one unittest to cover all, or at best short ddoc unittests showing the use of each inside a pure function